File: source/Python/iterator/hospital_1.py

import random
import string
import time

from hospital_base import HospitalBase
from patient_solution import Patient, time_to_mins


class Hospital_1(HospitalBase):

    def __init__(self):
        super().__init__()

        self.START_MINS = time_to_mins("08:00")
        self.END_MINS = time_to_mins("18:00")
        self.LUNCH_START_MINS = time_to_mins("12:00")
        self.LUNCH_END_MINS = time_to_mins("13:00")
        self.MINUTES_PER_SLOT = 20
        self.NUM_TIMESLOTS = (self.LUNCH_START_MINS - self.START_MINS + self.END_MINS - self.LUNCH_END_MINS)\
                             / self.MINUTES_PER_SLOT
        self.patients = [None] * int(self.NUM_TIMESLOTS)

    def __iter__(self):
        for el in self.patients:
            if el is not None:
                yield el

    def add_patient(self, patient: Patient):
        time_in_mins = time_to_mins(patient.time)
        if time_in_mins < self.START_MINS or time_in_mins >= self.END_MINS \
            or time_in_mins >= self.LUNCH_START_MINS and time_in_mins < self.LUNCH_END_MINS \
            or time_in_mins % self.MINUTES_PER_SLOT != 0:
            return False
        if time_in_mins < self.LUNCH_START_MINS:
            slot = int((time_in_mins - self.START_MINS) / self.MINUTES_PER_SLOT)
        else:
            numSlotsBeforeLunch = (self.LUNCH_START_MINS - self.START_MINS) / self.MINUTES_PER_SLOT
            slot = int((time_in_mins - self.LUNCH_END_MINS) / self.MINUTES_PER_SLOT + numSlotsBeforeLunch)

        if self.patients[slot] is not None:
            return False
        self.patients[slot] = patient
        return True


if __name__ == "__main__":

    """  ===================  Public tests   ===================  """
    ll = Hospital_1()
    ll.add_patient(Patient("Max", "11:00"))
    ll.add_patient(Patient("Alex", "13:00"))
    ll.add_patient(Patient("George", "14:00"))
    list_of_patients = [Patient("Max", "11:00"), Patient("Alex", "13:00"), Patient("George", "14:00")]
    for i, el in enumerate(ll):
        assert el == list_of_patients[i]

    """ ============= Hidden Tests ============= """
    # ================== One element only ==================
    ll = Hospital_1()
    ll.add_patient(Patient("Alex", "13:20"))
    for el in ll:
        assert el == Patient("Alex", "13:20")

    # ================== Incorrect time is not added ==================
    ll = Hospital_1()
    if_added = ll.add_patient(Patient("Alex", "13:25"))
    assert not if_added

    # ================== Do not add if the slot is occupied ==================
    ll = Hospital_1()
    ll.add_patient(Patient("Alex", "13:20"))
    if_added = ll.add_patient(Patient("Max", "13:20"))
    assert not if_added


    # ================== Normal Sort ==================
    ll = Hospital_1()
    ll.add_patient(Patient("Alex", "13:20"))
    ll.add_patient(Patient("Max", "11:00"))
    ll.add_patient(Patient("George", "14:00"))
    ll.add_patient(Patient("Lisa", "09:00"))
    list_of_patients = [Patient("Lisa", "09:00"), Patient("Max", "11:00"),
                        Patient("Alex", "13:20"), Patient("George", "14:00")]
    for i, el in enumerate(ll):
        assert el == list_of_patients[i]

    # ================== Same names ==================
    ll = Hospital_1()
    ll.add_patient(Patient("Alex", "13:20"))
    ll.add_patient(Patient("Alex", "11:00"))
    ll.add_patient(Patient("Alex", "14:00"))
    ll.add_patient(Patient("Alex", "09:00"))
    list_of_patients = [Patient("Alex", "09:00"), Patient("Alex", "11:00"),
                        Patient("Alex", "13:20"), Patient("Alex", "14:00")]
    for i, el in enumerate(ll):
        assert el == list_of_patients[i]

    # ================== Same times, different names ==================
    ll = Hospital_1()
    ll.add_patient(Patient("Daniel", "13:20"))
    ll.add_patient(Patient("Mark", "13:20"))
    ll.add_patient(Patient("Ben", "13:20"))
    ll.add_patient(Patient("Alex", "13:20"))
    list_of_patients = [Patient("Alex", "13:20"), Patient("Ben", "13:20"),
                        Patient("Daniel", "13:20"), Patient("Mark", "13:20")]
    for el in ll:
        assert el == Patient("Daniel", "13:20")

    # ================== Time limit test 1==================
    # Add in the beginning
    ll = Hospital_1()
    for hour in ["09", "10", "11", "13", "14", "15", "16"]:
        for minutes in ["00", "20", "40"]:
            patient_name = ''.join(random.choice(string.ascii_lowercase) for _ in range(8))
            patient_time = hour+":"+minutes
            ll.add_patient(Patient(patient_name, patient_time))
    t1 = time.time()
    ll.add_patient(Patient('aaa', "08:00"))
    t2 = time.time()
    assert t2-t1 < 0.00005

    # ================== Time limit test 2==================
    # Add in the end
    ll = Hospital_1()
    for hour in ["09", "10", "11", "13", "14", "15", "16"]:
        for minutes in ["00", "20", "40"]:
            patient_name = ''.join(random.choice(string.ascii_lowercase) for _ in range(8))
            patient_time = hour+":"+minutes
            ll.add_patient(Patient(patient_name, patient_time))
    t1 = time.time()
    ll.add_patient(Patient('zzz', "17:40"))
    t2 = time.time()
    assert t2-t1 < 0.00005

    # ================== Time limit test 3==================
    # Quick iterator
    ll = Hospital_1()
    for hour in ["09", "10", "11", "13", "14", "15", "16"]:
        for minutes in ["00", "20", "40"]:
            patient_name = ''.join(random.choice(string.ascii_lowercase) for _ in range(8))
            patient_time = hour+":"+minutes
            ll.add_patient(Patient(patient_name, patient_time))
    t1 = time.time()
    for el in ll:
        pass
    t2 = time.time()
    assert t2-t1 < 0.00005

File: source/JAVA/iterator/latest/Hospital1.java
import java.time.Duration;
import java.time.Instant;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;
import java.util.Random;
import java.util.stream.Stream;

public class Hospital1 extends HospitalBase {

    private static final int START_MINS = timeToMins("08:00");
    private static final int END_MINS = timeToMins("18:00");
    private static final int LUNCH_START_MINS = timeToMins("12:00");
    private static final int LUNCH_END_MINS = timeToMins("13:00");
    private static final int MINUTES_PER_SLOT = 20;
    private static final int NUM_TIMESLOTS = (LUNCH_START_MINS - START_MINS + END_MINS - LUNCH_END_MINS) / MINUTES_PER_SLOT;

    private final PatientBase[] patients;

    public Hospital1() {
        patients = new PatientBase[NUM_TIMESLOTS];
    }

    @Override
    public boolean addPatient(PatientBase patient) {
        String time = patient.getTime();
        int timeInMins = timeToMins(time);

        if (timeInMins < START_MINS
                || timeInMins >= END_MINS
                || timeInMins >= LUNCH_START_MINS && timeInMins < LUNCH_END_MINS
                || timeInMins % MINUTES_PER_SLOT != 0) {
            return false;
        }

        int slot;
        if (timeInMins < LUNCH_START_MINS) {
            slot = (timeInMins - START_MINS) / MINUTES_PER_SLOT;
        } else {
            int numSlotsBeforeLunch = (LUNCH_START_MINS - START_MINS) / MINUTES_PER_SLOT;
            slot = (timeInMins - LUNCH_END_MINS) / MINUTES_PER_SLOT + numSlotsBeforeLunch;
        }
        if (patients[slot] != null) {
            return false;
        }

        patients[slot] = patient;
        return true;
    }

    @Override
    public Iterator<PatientBase> iterator() {
        return Stream.of(patients).filter(Objects::nonNull).iterator();
    }

    private static int timeToMins(String time) {
        String[] components = time.split(":");
        return Integer.parseInt(components[0]) * 60 + Integer.parseInt(components[1]);
    }

    public static void main(String[] args) {
        /*
         * Public tests
         */
        {
            var hospital = new Hospital1();
            var p1 = new Patient("Max", "11:00");
            var p2 = new Patient("Alex", "13:00");
            var p3 = new Patient("George", "14:00");
            hospital.addPatient(p1);
            hospital.addPatient(p2);
            hospital.addPatient(p3);
            var patients = new Patient[] { p1, p2, p3 };
            int i = 0;
            for (var patient : hospital) {
                assert Objects.equals(patient, patients[i++]);
            }
        }

        /*
         * Hidden tests
         */
        // One element only
        {
            var hospital = new Hospital1();
            var p1 = new Patient("Alex", "13:20");
            hospital.addPatient(p1);
            for (var patient : hospital) {
                assert Objects.equals(patient, p1);
            }
        }

        // Incorrect time is not added
        {
            var hospital = new Hospital1();
            assert !hospital.addPatient(new Patient("Alex", "13:25"));
        }

        // Do not add if the slot is occupied
        {
            var hospital = new Hospital1();
            hospital.addPatient(new Patient("Alex", "13:20"));
            assert !hospital.addPatient(new Patient("Max", "13:20"));
        }

        // Normal sort
        {
            var hospital = new Hospital1();
            var p1 = new Patient("Alex", "13:20");
            var p2 = new Patient("Max", "11:00");
            var p3 = new Patient("George", "14:00");
            var p4 = new Patient("Lisa", "09:00");
            hospital.addPatient(p1);
            hospital.addPatient(p2);
            hospital.addPatient(p3);
            hospital.addPatient(p4);
            var patients = List.of(p4, p2, p1, p3);
            int i = 0;
            for (var patient : hospital) {
                assert Objects.equals(patient, patients.get(i++));
            }
        }

        // Same names
        {
            var hospital = new Hospital1();
            var p1 = new Patient("Alex", "13:20");
            var p2 = new Patient("Alex", "11:00");
            var p3 = new Patient("Alex", "14:00");
            var p4 = new Patient("Alex", "09:00");
            hospital.addPatient(p1);
            hospital.addPatient(p2);
            hospital.addPatient(p3);
            hospital.addPatient(p4);
            var patients = List.of(p4, p2, p1, p3);
            int i = 0;
            for (var patient : hospital) {
                assert Objects.equals(patient, patients.get(i++));
            }
        }

        // Same times, different names
        {
            var hospital = new Hospital1();
            var p1 = new Patient("Daniel", "13:20");
            var p2 = new Patient("Mark", "13:20");
            var p3 = new Patient("Ben", "13:20");
            var p4 = new Patient("Alex", "13:20");
            hospital.addPatient(p1);
            hospital.addPatient(p2);
            hospital.addPatient(p3);
            hospital.addPatient(p4);
            for (var patient : hospital) {
                assert Objects.equals(patient, p1);
            }
        }

        // Time limit test 1 (add in the beginning)
        {
            var hospital = new Hospital1();
            var random = new Random();
            for (String hour : List.of("09", "10", "11", "13", "14", "15", "16")) {
                for (String minutes : List.of("00", "20", "40")) {
                    String patientName = random.ints(8, 'a', 'z' + 1)
                            .collect(StringBuffer::new, StringBuffer::append, StringBuffer::append)
                            .toString();
                    String patientTime = String.format("%s:%s", hour, minutes);
                    hospital.addPatient(new Patient(patientName, patientTime));
                }
            }
            var t1 = Instant.now();
            hospital.addPatient(new Patient("aaa", "08:00"));
            var t2 = Instant.now();
            assert Duration.between(t1, t2).toNanos() < 50_000;
        }

        // Time limit test 1 (add in the end)
        {
            var hospital = new Hospital1();
            var random = new Random();
            for (String hour : List.of("09", "10", "11", "13", "14", "15", "16")) {
                for (String minutes : List.of("00", "20", "40")) {
                    String patientName = random.ints(8, 'a', 'z' + 1)
                            .collect(StringBuffer::new, StringBuffer::append, StringBuffer::append)
                            .toString();
                    String patientTime = String.format("%s:%s", hour, minutes);
                    hospital.addPatient(new Patient(patientName, patientTime));
                }
            }
            var t1 = Instant.now();
            hospital.addPatient(new Patient("zzz", "17:40"));
            var t2 = Instant.now();
            assert Duration.between(t1, t2).toNanos() < 50_000;
        }

        // Time limit test 1 (quick iterator)
        {
            var hospital = new Hospital1();
            var random = new Random();
            for (String hour : List.of("09", "10", "11", "13", "14", "15", "16")) {
                for (String minutes : List.of("00", "20", "40")) {
                    String patientName = random.ints(8, 'a', 'z' + 1)
                            .collect(StringBuffer::new, StringBuffer::append, StringBuffer::append)
                            .toString();
                    String patientTime = String.format("%s:%s", hour, minutes);
                    hospital.addPatient(new Patient(patientName, patientTime));
                }
            }
            var t1 = Instant.now();
            for (var patient : hospital) {
            }
            var t2 = Instant.now();
            assert Duration.between(t1, t2).toNanos() < 50_000;
        }
    }
}
